### PR TITLE
EMAUS (Ensemble Microcanonical Adjusted-Unadjusted Sampler)

### PR DESCRIPTION
**Pull request by reubenharry**
This is in essence a tuning scheme for running multiple chains in parallel (ensemble). The key trick is to use MCLMC without MH (i.e. unadjusted) first in order to very quickly find the typical set, and then to use MH for the steps you actually count. The details are in how to decide when to switch between these two phases, and how to tune the hyperparameters using estimates from the ensemble of chains.

The upshot is that the method is fast, even compared to other ensemble methods.

Changes:
- a new ensemble_mclmc: TODO
- util.py: TODO
- other: TODO

 A few important guidelines and requirements before we can merge your PR:

 - [X] We should be able to understand what the PR does from its title only;
 - [X] There is a high-level description of the changes;
 - [ ] There are links to *all* the relevant issues, discussions and PRs;
 - [ ] The branch is rebased on the latest `main` commit;
 - [ ] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [ ] The code respects the current naming conventions;
 - [ ] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [ ] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [ ] There are tests covering the changes;
 - [ ] The doc is up-to-date;
 - [ ] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.

### CODE REVIEW

**Comment by reubenharry** [comment]
I unmarked this as a draft, even though there's still some rough edges, just in case @junpenglao or anyone else wanted to give high level feedback over the main changes (e.g. should I rearrange the files and put everything in one file, etc).

I see I have also somehow added spaces to a bunch of files, not sure how this happened or how to fix...

**Comment by reubenharry** [comment]
This should be ready to review @junpenglao . I realize it's a large PR, so let me know if anything can be done to make it easier.

Confusingly, the new test passes locally, but fails on CI. I don't understand what has changed (the random seed seems to be the same).

**Comment by junpenglao** [comment]
Great work, let me take a look this week.

**Comment by reubenharry** [comment]
> > I don't understand what has changed (the random seed seems to be the same).
> 
> The latest release of JAX seems to change some of the seeding output, try either a new random key or in crease the number of samples.

Resolved, thanks!

File: blackjax/adaptation/step_size.py [hunk 1]
```diff
@@ -257,3 +257,45 @@ def update(rss_state: ReasonableStepSizeState) -> ReasonableStepSizeState:
...
+
+        stepsize = terminated * exp_x + (1 - terminated) * jnp.exp(x_new)
+
+        terminated_new = (
+            jnp.abs(acc_rate_new - acc_prob_wanted) < tolerance
+        ) | terminated
+
+        return (bounds_new, terminated_new), stepsize
+
+    return (jnp.array([-jnp.inf, jnp.inf]), False), update
```
Comment (junpenglao) [inline blackjax/adaptation/step_size.py:301]: Is it really necessary to return `(jnp.array([-jnp.inf, jnp.inf]), False)`? They are static.

File: blackjax/adaptation/step_size.py [hunk 2]
```diff
@@ -257,3 +257,45 @@ def update(rss_state: ReasonableStepSizeState) -> ReasonableStepSizeState:
...
+        acc_high = acc_rate_new > acc_prob_wanted
+        x = jnp.log(exp_x)
+
+        def on_true(bounds):
+            bounds0 = jnp.max(jnp.array([bounds[0], x]))
+            return jnp.array([bounds0, bounds[1]]), bounds0 + reduce_shift
+
+        def on_false(bounds):
+            bounds1 = jnp.min(jnp.array([bounds[1], x]))
+            return jnp.array([bounds[0], bounds1]), bounds1 - reduce_shift
```
Comment (junpenglao) [inline blackjax/adaptation/step_size.py:278]:
```suggestion
        def on_true(bounds):
            lower, upper = bounds
            lower = jnp.max(jnp.array([lower, x]))
            return jnp.array([lower, upper]), lower + reduce_shift

        def on_false(bounds):
            lower, upper = bounds
            upper = jnp.min(jnp.array([upper, x]))
            return jnp.array([lower, upper]), upper - reduce_shift
```

**Review by junpenglao — COMMENTED** [review]
I am still reviewing the main functions.

**Comment by reubenharry** [comment]
OK, tests are up and running again! @junpenglao
